### PR TITLE
Adds fast gradient clipping support for the Embedding layer.

### PR DESCRIPTION
**Pull request by pagarwl**
The algorithm used is described in the 'A Unified Fast Gradient Clipping Framework for DP-SGD' paper: https://proceedings.neurips.cc/paper_files/paper/2023/file/a45d344b28179c8da7646bc38ff50ad8-Paper-Conference.pdf. 
 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

Previously, Ghost clipping was not supported in Opacus for embedding layer. With default DP-SGD implementation, the training OOMs out on large embedding layers over large physical batch size (useful for privacy). The regular DP-SGD needs O(Bnd), where B=physical batch size, n=vocab size, d=embedding dimension. To give an example on memory needed: we've seen embeddings with [vocab size=1000000, dim=5] (and higher) in real-world differential privacy applications. With a physical batch size of 16,000, memory needed: 16000 × 1000000 × 5 x 4 = 298.02 GB. 

With this change, we need significantly smaller memory: O(Br) where B is physical batch size, and r is number of unique  indices in the embedding sequence. We could successfully run DP-SGD over above example, using < 8GiB.

This is a good add to Opacus, enabling larger embedding layers over larger physical batch sizes to be trained with DP-SGD.

## How Has This Been Tested (if it applies)

Unit tests. Runs over large embedding layer training over realworld DP application.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by EnayatUllah** [comment]
Thank you for the PR! Overall, the logic look good to me -- I left some minor comments. Additionally, I suggest to merge the content of the file `embedding_norm_sample.py` to `embedding.py` itself.

File: opacus/tests/grad_sample_module_fast_gradient_clipping_test.py [hunk 1]
```diff
@@ -260,3 +276,151 @@ def test_gradient_calculation_fast_gradient_clipping(self, size, length, dim):
...
+        grad_sample_module.disable_hooks()
+        second_loss.backward()
+
+        all_norms_gc = [param._norm_sample for param in grad_sample_module.parameters()]
+        flat_norms_gc = torch.cat([p.flatten() for p in all_norms_gc])
+
+        diff = flat_norms_normal - flat_norms_gc
+
+        logging.info(f"Diff = {diff}")
+        msg = "Fail: Gradients from vanilla DP-SGD and from fast gradient clipping are different"
```
Comment (EnayatUllah) [inline opacus/tests/grad_sample_module_fast_gradient_clipping_test.py:354]: msg should be "Fail: Gradient norms from vanilla DP-SGD and from fast gradient clipping are different"

File: opacus/tests/grad_samples/embedding_norm_sample_test.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        # Call the function we want to test
+        result = embedding_norm_sample.compute_embedding_norm_sample(
+            embedding_layer, activations, backprops
+        )
+
+        # Expected output based on the example
+        expected_norms = torch.tensor(
+            [0.0150, 0.0071, 0.0005, 0.0081, 0.0039], dtype=torch.float32
+        )
+        print("expected_norms: ", expected_norms)
```
Comment (EnayatUllah) [inline opacus/tests/grad_samples/embedding_norm_sample_test.py:214]: Remove the print statement

File: opacus/grad_sample/embedding_norm_sample.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    """
+    device = activations[0].device
+    input_ids = activations[0].to(device)
+    grad_values = backprops.to(device)
+
+    # Reshape input_ids preserving the batch size as the first dimension
+    input_ids = input_ids.reshape(input_ids.shape[0], -1)
+
+    # Reshape grad_values preserving the embedding dimension as the last dimension
+    grad_values = grad_values.reshape(-1, grad_values.size(-1))
```
Comment (HuanyuZhang) [inline opacus/grad_sample/embedding_norm_sample.py:113]: @pagarwl  just double checking, this implementation not only works for 2 dimensions ([batch_size, seq_length]), but also works for any general dimension, correct?

File: opacus/tests/grad_samples/embedding_norm_sample_test.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    def test_compute_embedding_norm_sample_with_non_one_embedding_dim(self):
+        # Define the embedding layer
+        embedding_dim = 2
+        vocab_size = 3
+        embedding_layer = nn.Embedding(vocab_size, embedding_dim)
+
+        # Manually set weights for the embedding layer for testing
+        embedding_layer.weight = nn.Parameter(
+            torch.tensor([[0.1], [0.2], [0.3]], dtype=torch.float32)
+        )
```
Comment (HuanyuZhang) [inline opacus/tests/grad_samples/embedding_norm_sample_test.py:74]: Nitpicking: although it will not be used in calculation, could we change the shape from [3,1] to [3,2] to match the embedding_dim=2? @pagarwl

File: opacus/tests/grad_samples/embedding_norm_sample_test.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+        # Example input ids (activations). Shape: [6, 1, 1].
+        input_ids = torch.tensor(
+            [[[1]], [[1]], [[2]], [[0]], [[2]], [[0]]], dtype=torch.long
+        )
+
+        # Example gradients per input id, with embedding_dim=2.
+        # Shape: [6, 1, 1, 2]
```
Comment (HuanyuZhang) [inline opacus/tests/grad_samples/embedding_norm_sample_test.py:82]: It seems the shape is in a different logic from the first example (``test_compute_embedding_norm_sample``), where input has shape [3,2] and ``grad_values`` has shape [6,1]. Do you mind clarifying?

File: opacus/tests/grad_samples/embedding_norm_sample_test.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        # Manually set weights for the embedding layer for testing
+        embedding_layer.weight = nn.Parameter(
+            torch.tensor([[0.1], [0.2], [0.3]], dtype=torch.float32)
+        )
+
+        # Example input ids (activations). Shape: [3, 2]
+        input_ids = torch.tensor([[1, 1], [2, 0], [2, 0]], dtype=torch.long)
+
+        # Example gradients with respect to the embedding output (backprops).
+        # Shape: [6, 1]
```
Comment (HuanyuZhang) [inline opacus/tests/grad_samples/embedding_norm_sample_test.py:40]: Nitpicking: we should have size [3,2,1] for backprops, correct? [6, 1] is for grad_values.

File: opacus/tests/grad_samples/embedding_norm_sample_test.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        )
+
+        # Example input ids (activations). Shape: [6, 1, 1].
+        input_ids = torch.tensor(
+            [[[1]], [[1]], [[2]], [[0]], [[2]], [[0]]], dtype=torch.long
+        )
+
+        # Example gradients per input id, with embedding_dim=2.
+        # Shape: [6, 1, 1, 2]
+        grad_values = torch.tensor(
```
Comment (HuanyuZhang) [inline opacus/tests/grad_samples/embedding_norm_sample_test.py:83]: Similar here, maybe we should distinguish between grad_values and backprops, since they have different shapes?

File: opacus/grad_sample/embedding_norm_sample.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+      A dictionary of parameter gradients
+
+    NOTE: Here is an example input, and the expected intermediate values. This
+    is proivided to help in understanding the algorithm:
+    Inputs:
+      layer:  Embedding(3, 1) # (vocab_size, embedding_dim)
+      activations:  [tensor([[1, 1],
+          [2, 0],
+          [2, 0]])]
+      backprops:  tensor([[0.2000],
```
Comment (HuanyuZhang) [inline opacus/grad_sample/embedding_norm_sample.py:49]: Wrong shape for backprops, which should be [3,2,1]. [6,1] is for grad_values

**Review by HuanyuZhang — COMMENTED** [review]
Do you mind also fixing the lint error due to isort? https://github.com/pytorch/opacus/actions/runs/12342666465/job/34611577396

File: opacus/tests/grad_samples/embedding_norm_sample_test.py (hunk 4 above)
Comment (pagarwl) [reply opacus/tests/grad_samples/embedding_norm_sample_test.py:70]: Done.

File: opacus/tests/grad_sample_module_fast_gradient_clipping_test.py (hunk 1 above)
Comment (pagarwl) [reply opacus/tests/grad_sample_module_fast_gradient_clipping_test.py]: Done.

File: opacus/tests/grad_samples/embedding_norm_sample_test.py (hunk 2 above)
Comment (pagarwl) [reply opacus/tests/grad_samples/embedding_norm_sample_test.py]: Done.

File: opacus/tests/grad_samples/embedding_norm_sample_test.py (hunk 6 above)
Comment (pagarwl) [reply opacus/tests/grad_samples/embedding_norm_sample_test.py]: yes, updated.

File: opacus/tests/grad_samples/embedding_norm_sample_test.py (hunk 7 above)
Comment (pagarwl) [reply opacus/tests/grad_samples/embedding_norm_sample_test.py]: Done.

File: opacus/grad_sample/embedding_norm_sample.py (hunk 8 above)
Comment (pagarwl) [reply opacus/grad_sample/embedding_norm_sample.py]: Done.

**Comment by pagarwl** [comment]
> Do you mind also fixing the lint error due to isort? https://github.com/pytorch/opacus/actions/runs/12342666465/job/34611577396

Done.

**Comment by pagarwl** [comment]
> Thank you for the PR! Overall, the implementation looks good to me -- I left some minor comments. Additionally, I suggest to merge the content of the file `embedding_norm_sample.py` to `embedding.py` itself.

Thanks for the review. I have incorporated all the changes. My colleague can pursue merging the contents of`embedding_norm_sample.py` to `embedding.py` in a follow-up PR. I've moved to another team in the last couple of months and I am pressed on time.

**Comment by EnayatUllah** [comment]
Looks good to me. @pagarwl, can you fix the isort error? Run

```
isort -l 88 -o opacus --lines-after-imports 2 -m 3 --trailing-comma  .
```
from the repository root. Configuration for isort is located in .isort.cfg.
If using `isort` versions `<5.0.0` call
```bash
isort -l 88 -o opacus --lines-after-imports 2 -m 3 --trailing-comma --recursive
